iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/CompoundTextRange.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.tree.impl;
+
+import java.util.List;
+import org.sonar.iac.common.api.tree.impl.TextRange;
+import org.sonar.iac.common.api.tree.impl.TextRanges;
+
+public class CompoundTextRange extends TextRange {
+  private final List<TextRange> textRanges;
+
+  public CompoundTextRange(List<TextRange> textRanges) {
+    super(textRanges.get(0).start(), textRanges.get(textRanges.size() - 1).end());
+    this.textRanges = textRanges;
+  }
+
+  /**
+   * Provide the correct {@link TextRange}, which can extend a multiple lines.
+   * It combines the {@code startIndex} and the {@code token} length to find which of the {@link #textRanges} are covered, to compute the correct single/multi line {@link TextRange}.
+   */
+  public TextRange computeTextRangeAtIndex(int startIndex, String token) {
```

Comment:
It's complex logic. After you add tests try to refactor it. Suggested algorithm:
TextRange startRange = findStartRange(...)
// calculate `i` based on `startRange`
TextRange endRange = findEndRange(...)
// calculate the rest

I know come calculation will be done twice, but it should be easier to understand it.

Reply:
I refactored as much as I could.
I get rid of the `i` variable by using an Iterator instead.

I identified the elements that I need to keep in the context: the iterator's current value and the index which changes all along.
For the Iterator's current value, I added a wrapper for it.
For the index value, we keep it through the new method `navigateToRangeAtIndex` return value.

I also added some minor improvements.